devito/operator.py

```diff
@@ -117,24 +119,66 @@ def __init__(self, expressions, **kwargs):
         # Introduce the required symbol declarations
         nodes = iet_insert_C_decls(dle_state.nodes, self.func_table)
 
-        # Initialise ArgumentEngine
-        self.argument_engine = ArgumentEngine(clusters.ispace, parameters,
-                                              self.dle_arguments)
+        # Insert data and pointer casts for array parameters and profiling structs
+        functions = FindSymbols('symbolics').visit(nodes)
+        casts = [ArrayCast(f) for f in functions if f.is_Tensor and f._mem_external]
+        profiler = Object(self.profiler.name, self.profiler.dtype, self.profiler.new)
+        casts.append(PointerCast(profiler))
+        nodes = (List(body=casts), nodes)
```

Comment:
Hmm, I actually tried generalising this and moving it to a utility function, because the same thing needs to be done for newly created elementals. However, this is not as easy as it looks, due to the funky way the profiler `Object`s are handled. Since elementals don't need profiling, I think it would be great to simply pick the `Object` off the `TimedList` IET nodes, but that soon became messy. I'll give this another go, but maybe in a separate PR?